Descriptor method for *dynamic.Message does not work

Hey! I love this library! Thank you so much for its existence!

I'm having a strange result calling `desc.LoadMessageDescriptorForMessage` on a dynamic message with a `MessageDescriptor` that came from `builder`. Here's a reproducible test case:

```go
package prtc

import (
	"github.com/jhump/protoreflect/desc"
	"github.com/jhump/protoreflect/desc/builder"
	"github.com/jhump/protoreflect/dynamic"
	"testing"
)

func TestDescriptorGet(t *testing.T) {
	md, err := builder.NewMessage("Person").
		AddField(builder.NewField("name", builder.FieldTypeString())).
		Build()

	if err != nil {
		t.Fatalf("error building Person: %v", err)
	}

	msg := dynamic.NewMessage(md)
	msg.SetFieldByNumber(1, "Cailin")

	_, err = desc.LoadMessageDescriptorForMessage(msg)
	if err != nil {
		t.Fatalf("error loading Person's message descriptor: %v", err)
	}
}
```

And the output:

```console
$ go test
--- FAIL: TestDescriptorGet (0.00s)
panic: Failed to get descriptor path for Person [recovered]
        panic: Failed to get descriptor path for Person

goroutine 5 [running]:
testing.tRunner.func1(0xc4200c00f0)
        /usr/local/opt/go/libexec/src/testing/testing.go:742 +0x29d
panic(0x11aa580, 0xc42000e650)
        /usr/local/opt/go/libexec/src/runtime/panic.go:505 +0x229
github.com/kivikakk/prtc/vendor/github.com/jhump/protoreflect/dynamic.(*Message).Descriptor(0xc420088d80, 0xc420088660, 0x11f9d07, 0x6, 0x13320e0, 0x11f8de0, 0xc400000000)
        /Users/kivikakk/.go/src/github.com/kivikakk/prtc/vendor/github.com/jhump/protoreflect/dynamic/dynamic_message.go:176 +0x7bc
github.com/kivikakk/prtc/vendor/github.com/jhump/protoreflect/desc.loadMessageDescriptorForTypeLocked(0x11f9d07, 0x6, 0x13f3020, 0xc420088d80, 0xc420088d80, 0x0, 0x0)
        /Users/kivikakk/.go/src/github.com/kivikakk/prtc/vendor/github.com/jhump/protoreflect/desc/descriptor.go:1785 +0x76
github.com/kivikakk/prtc/vendor/github.com/jhump/protoreflect/desc.LoadMessageDescriptorForMessage(0x1221e00, 0xc420088d80, 0x0, 0x0, 0x0)
        /Users/kivikakk/.go/src/github.com/kivikakk/prtc/vendor/github.com/jhump/protoreflect/desc/descriptor.go:1765 +0x10b
github.com/kivikakk/prtc.TestDescriptorGet(0xc4200c00f0)
        /Users/kivikakk/.go/src/github.com/kivikakk/prtc/main_test.go:22 +0x18f
testing.tRunner(0xc4200c00f0, 0x12057e0)
        /usr/local/opt/go/libexec/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
        /usr/local/opt/go/libexec/src/testing/testing.go:824 +0x2e0
exit status 2
FAIL    github.com/kivikakk/prtc        0.010s
```

You can fetch this from [`github.com/kivikakk/prtc`](https://github.com/kivikakk/prtc) if it makes life easier.

I tried hacking the code in `dynamic_message.go` a bit to see what the deal is — this case here isn't finding anything:

https://github.com/jhump/protoreflect/blob/6f4f3bba11fd817df7f57c9b319eeb55a2590167/dynamic/dynamic_message.go#L167-L173

This is because `Person` has no nested message types, but we start our search at `m.md` i.e. `Person` itself, and don't match anything with that name. I tried starting the search from `m.md.GetFile()` instead, just to see if it'd work, and it sure didn't; the path code presumes there'll be at least 2 items. And when I changed it to not presume that, … actually, wait a second:

https://github.com/jhump/protoreflect/blob/6f4f3bba11fd817df7f57c9b319eeb55a2590167/dynamic/dynamic_message.go#L180-L186

`j` starts off as `len(path)`. We try to read `path[j]`. Won't that always fail, as it'll be one past the end of the slice?

With the following diff applied, I get success at loading descriptors of dynamic messages with non-`builder`-origin types:

```diff
diff --git a/dynamic/dynamic_message.go b/dynamic/dynamic_message.go
index 37ab9ae..81ad546 100644
--- a/dynamic/dynamic_message.go
+++ b/dynamic/dynamic_message.go
@@ -154,7 +154,7 @@ func (m *Message) Descriptor() ([]byte, []int) {
        path := []int{}
        var d desc.Descriptor
        name := m.md.GetFullyQualifiedName()
-       for d = m.md; d != nil; d = d.GetParent() {
+       for d = m.md.GetParent(); d != nil; d = d.GetParent() {
                found := false
                switch d := d.(type) {
                case (*desc.FileDescriptor):
@@ -178,7 +178,7 @@ func (m *Message) Descriptor() ([]byte, []int) {
        }
        // reverse the path
        i := 0
-       j := len(path)
+       j := len(path) - 1
        for i < j {
                path[i], path[j] = path[j], path[i]
                i++
```

And the following failure for the test case in the issue body:

```console
$ go test
--- FAIL: TestBuiltMessage (0.00s)
        dynamic_message_test.go:1970: error loading Person's message descriptor: failed to decompress "Person" descriptor: bad gzipped descriptor: unexpected EOF
FAIL
exit status 1
FAIL    github.com/jhump/protoreflect/dynamic   0.147s
```

--- a/dynamic/dynamic_message.go
+++ b/dynamic/dynamic_message.go
@@ -158,7 +158,7 @@ func (m *Message) Descriptor() ([]byte, []int) {
 	path := []int{}
 	var d desc.Descriptor
 	name := m.md.GetFullyQualifiedName()
-	for d = m.md.GetParent(); d != nil; d = d.GetParent() {
+	for d = m.md.GetParent(); d != nil; name, d = d.GetFullyQualifiedName(), d.GetParent() {
 		found := false
 		switch d := d.(type) {
 		case (*desc.FileDescriptor):
@@ -166,9 +166,6 @@ func (m *Message) Descriptor() ([]byte, []int) {
 				if md.GetFullyQualifiedName() == name {
 					found = true
 					path = append(path, i)
-					if p := md.GetParent(); p != nil {
-						name = p.GetFullyQualifiedName()
-					}
 				}
 			}
 		case (*desc.MessageDescriptor):
@@ -176,9 +173,6 @@ func (m *Message) Descriptor() ([]byte, []int) {
 				if md.GetFullyQualifiedName() == name {
 					found = true
 					path = append(path, i)
-					if p := md.GetParent(); p != nil {
-						name = p.GetFullyQualifiedName()
-					}
 				}
 			}
 		}

--- a/dynamic/dynamic_message.go
+++ b/dynamic/dynamic_message.go
@@ -166,6 +166,9 @@ func (m *Message) Descriptor() ([]byte, []int) {
 				if md.GetFullyQualifiedName() == name {
 					found = true
 					path = append(path, i)
+					if p := md.GetParent(); p != nil {
+						name = p.GetFullyQualifiedName()
+					}
 				}
 			}
 		case (*desc.MessageDescriptor):
@@ -173,6 +176,9 @@ func (m *Message) Descriptor() ([]byte, []int) {
 				if md.GetFullyQualifiedName() == name {
 					found = true
 					path = append(path, i)
+					if p := md.GetParent(); p != nil {
+						name = p.GetFullyQualifiedName()
+					}
 				}
 			}
 		}

--- a/dynamic/dynamic_message.go
+++ b/dynamic/dynamic_message.go
@@ -149,12 +149,16 @@ func (m *Message) Descriptor() ([]byte, []int) {
 	if err != nil {
 		panic(fmt.Sprintf("Failed to get encoded descriptor for %s", m.md.GetFile().GetName()))
 	}
+	err = w.Close()
+	if err != nil {
+		panic(fmt.Sprintf("Failed to get an encoded descriptor for %s", m.md.GetFile().GetName()))
+	}
 
 	// and path to message
 	path := []int{}
 	var d desc.Descriptor
 	name := m.md.GetFullyQualifiedName()
-	for d = m.md; d != nil; d = d.GetParent() {
+	for d = m.md.GetParent(); d != nil; d = d.GetParent() {
 		found := false
 		switch d := d.(type) {
 		case (*desc.FileDescriptor):
@@ -178,7 +182,7 @@ func (m *Message) Descriptor() ([]byte, []int) {
 	}
 	// reverse the path
 	i := 0
-	j := len(path)
+	j := len(path) - 1
 	for i < j {
 		path[i], path[j] = path[j], path[i]
 		i++

--- a/dynamic/dynamic_message_test.go
+++ b/dynamic/dynamic_message_test.go
@@ -9,6 +9,7 @@ import (
 
 	"github.com/jhump/protoreflect/desc"
 	"github.com/jhump/protoreflect/desc/builder"
+	"github.com/jhump/protoreflect/internal"
 	"github.com/jhump/protoreflect/internal/testprotos"
 	"github.com/jhump/protoreflect/internal/testutil"
 )
@@ -1957,18 +1958,17 @@ func TestBuiltMessage(t *testing.T) {
 	md, err := builder.NewMessage("Person").
 		AddField(builder.NewField("name", builder.FieldTypeString())).
 		Build()
-
-	if err != nil {
-		t.Fatalf("error building Person: %v", err)
-	}
+	testutil.Ok(t, err, "failed to build message")
 
 	msg := NewMessage(md)
-	msg.SetFieldByNumber(1, "Cailin")
+	gzipped, actualPath := msg.Descriptor()
+	_, expectedPath := (*testprotos.TestMessage)(nil).Descriptor()
+	testutil.Eq(t, expectedPath, actualPath, "descriptor paths are not the same")
 
-	_, err = desc.LoadMessageDescriptorForMessage(msg)
-	if err != nil {
-		t.Fatalf("error loading Person's message descriptor: %v", err)
-	}
+	actualFd, err := internal.DecodeFileDescriptor("test", gzipped)
+	testutil.Ok(t, err, "failed to decode gzipped descriptor")
+
+	testutil.Ceq(t, md.GetFile().AsFileDescriptorProto(), actualFd, eqpm, "file descriptors are not the same")
 }
 
 func TestBuiltNestedMessage(t *testing.T) {
@@ -1979,20 +1979,18 @@ func TestBuiltNestedMessage(t *testing.T) {
 		AddField(builder.NewField("name", builder.FieldTypeString())).
 		AddNestedMessage(submb).
 		Build()
-
-	if err != nil {
-		t.Fatalf("error building Person: %v", err)
-	}
+	testutil.Ok(t, err, "failed to build message")
 
 	nmt := md.GetNestedMessageTypes()[0]
-
 	msg := NewMessage(nmt)
-	msg.SetFieldByNumber(1, uint32(40))
+	gzipped, actualPath := msg.Descriptor()
+	_, expectedPath := (*testprotos.TestMessage_NestedMessage)(nil).Descriptor()
+	testutil.Eq(t, expectedPath, actualPath, "descriptor paths are not the same")
 
-	_, err = desc.LoadMessageDescriptorForMessage(msg)
-	if err != nil {
-		t.Fatalf("error loading Person.Limb's message descriptor: %v", err)
-	}
+	actualFd, err := internal.DecodeFileDescriptor("test", gzipped)
+	testutil.Ok(t, err, "failed to decode gzipped descriptor")
+
+	testutil.Ceq(t, md.GetFile().AsFileDescriptorProto(), actualFd, eqpm, "file descriptors are not the same")
 }
 
 type panicError struct {

--- a/dynamic/dynamic_message_test.go
+++ b/dynamic/dynamic_message_test.go
@@ -1971,6 +1971,30 @@ func TestBuiltMessage(t *testing.T) {
 	}
 }
 
+func TestBuiltNestedMessage(t *testing.T) {
+	submb := builder.NewMessage("Limb").
+		AddField(builder.NewField("length", builder.FieldTypeUInt32()))
+
+	md, err := builder.NewMessage("Person").
+		AddField(builder.NewField("name", builder.FieldTypeString())).
+		AddNestedMessage(submb).
+		Build()
+
+	if err != nil {
+		t.Fatalf("error building Person: %v", err)
+	}
+
+	nmt := md.GetNestedMessageTypes()[0]
+
+	msg := NewMessage(nmt)
+	msg.SetFieldByNumber(1, uint32(40))
+
+	_, err = desc.LoadMessageDescriptorForMessage(msg)
+	if err != nil {
+		t.Fatalf("error loading Person.Limb's message descriptor: %v", err)
+	}
+}
+
 type panicError struct {
 	panic interface{}
 }

--- a/dynamic/dynamic_message_test.go
+++ b/dynamic/dynamic_message_test.go
@@ -8,6 +8,7 @@ import (
 	"github.com/golang/protobuf/proto"
 
 	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/desc/builder"
 	"github.com/jhump/protoreflect/internal/testprotos"
 	"github.com/jhump/protoreflect/internal/testutil"
 )
@@ -1950,6 +1951,24 @@ func TestMergeInto(t *testing.T) {
 
 func TestMergeFrom(t *testing.T) {
 	// TODO
+}
+
+func TestBuiltMessage(t *testing.T) {
+	md, err := builder.NewMessage("Person").
+		AddField(builder.NewField("name", builder.FieldTypeString())).
+		Build()
+
+	if err != nil {
+		t.Fatalf("error building Person: %v", err)
+	}
+
+	msg := NewMessage(md)
+	msg.SetFieldByNumber(1, "Cailin")
+
+	_, err = desc.LoadMessageDescriptorForMessage(msg)
+	if err != nil {
+		t.Fatalf("error loading Person's message descriptor: %v", err)
+	}
 }
 
 type panicError struct {

